File: remove-duplicates-from-sorted-list/solution83.go
// https://leetcode.com/problems/remove-duplicates-from-sorted-list/

package main

import (
	"fmt"
	"strconv"
	"strings"
)

type ListNode struct {
	Val  int
	next *ListNode
}

// T: O(n), S: O(1)
func deleteDuplicates(head *ListNode) *ListNode {
	curr := head
	for curr != nil && curr.next != nil {
		if curr.next.Val == curr.Val {
			curr.next = curr.next.next
		} else {
			curr = curr.next
		}
	}
	return head
}

func (node *ListNode) String() string {
	var sb strings.Builder
	sb.WriteString("[ ")
	curr := node
	for curr != nil {
		sb.WriteString(strconv.Itoa(curr.Val))
		sb.WriteString(" ")
		curr = curr.next
	}
	sb.WriteString("]")
	return sb.String()
}

func main() {
	// [1,1,2]
	head := &ListNode{Val: 1}
	head.next = &ListNode{Val: 1}
	head.next.next = &ListNode{Val: 2}
	fmt.Println(head.String()) // [1,1,2]
	head = deleteDuplicates(head)
	fmt.Println(head.String()) // [1,2]
}

File: remove-duplicates-from-sorted-list/Solution83.java
// https://leetcode.com/problems/remove-duplicates-from-sorted-list/

class Solution83 {

    // T: O(n), S: O(1)
    public ListNode deleteDuplicates(ListNode head) {
        var curr = head;
        while (curr != null && curr.next != null) {
            if (curr.next.val == curr.val) {
                curr.next = curr.next.next;
            } else {
                curr = curr.next;
            }
        }
        return head;
    }

    public static void main(String[] args) {
        var sol = new Solution83();
        // [1,1,2]
        var head = new ListNode(1);
        head.next = new ListNode(1);
        head.next.next = new ListNode(2);
        System.out.println(head); // 1 -> 1 -> 2
        sol.deleteDuplicates(head);
        System.out.println(head);
    }

}
